### DISCUSSION TITLE
ELisp Interpretation

### DISCUSSION THREAD

**Discussion by skeledrew** [hylang/hy#2344] [Q&A] [answered]
Hy! I'm working on a project into which I'd like to interpret ELisp code, and I'm thinking to use Hy as the base. How feasible would this be, particularly in changing the behavior of Hy's Symbol to mimic that of Emacs', as well as `obarray`-style namespacing? What other Hy-specific considerations do I need to keep in mind?

**Comment by Kodiologist** [comment, marked as answer]
I'm afraid you have your work cut out for you. The semantic differences between Emacs Lisp and Hy run deep, and a lot of the details of Hy's implementation reflect Python semantics and Python's `ast` structure. You may have better luck writing an Emacs Lisp interpreter from scratch (perhaps in Hy) than using Hy's codebase. Remember that Hy's code is concerned only with translating Hy into Python `ast` objects rather than full-blow interpretation, which is probably what you'd need to do here.

**Reply by skeledrew** [reply]
Could you list some of the more challenging differences? That'd help me decide if it's worth modifying Hy or just going from scratch as you recommend. But ideally I really would prefer things to lean more toward the Python side, re semantics and ast. I'm hoping for a solution that allows for as-close-as-practical feature parity with ELisp, but is fully usable from within Python, ie. an el -> py(c) transpiler.

**Reply by Kodiologist** [reply]
It's been a long time since I had to think about the details of Elisp, but the first thing that comes to mind is that binding is dynamic by default, and Python doesn't really have dynamic binding. You'll have to simulate it with globals. The second is that Elisp is written with the expectation that text-editor-like constructs like buffers are convenient and so it approaches text processing in a way that looks weird in any other programming language. You'll have to implement a lot of text-editor-like features despite that fact that you're not writing a text editor. There are a lot of small things like how Elisp uses machine words for its integer type whereas Python uses arbitrary-size integers.

**Reply by skeledrew** [reply]
I see. Actually a text editor is part of my project, and as I've been using Emacs for a few years and like how it works, I'm hoping to replicate the functionality at least partially, but also natively provide access to the Python ecosystem. Thank you!
